### PR TITLE
Rigid Origami Folding

### PR DESCRIPTION
**Pull request by asahidari**
## Addressed problem description

This PR enable us to simulate wide range of rigid origami fold. Previously, I added #3375 for only one type of folding, but this PR can offer various type of origami folding. Also, this is related to #1381 'WISHLIST' of origami fold.

## Solution description

I added a script node to SNLite_templates/utils directory to avoid any influence or changes to other modules.
Inputs and outputs of this script node are following. I see this link as a reference: https://www.grasshopper3d.com/forum/topics/new-origami-example?groupUrl=kangaroo&groupId=2985220%3AGroup%3A120977&id=2985220%3ATopic%3A973616&page=3

- [in] obj_in (object)
- [in] valleys (edge list)
- [in] valley angle (scalar value)
- [in] mountains (edge list)
- [in] mountain angle (scalar value)
- [in] folding (scalar value, ratio of folding)
- [in] step (scalar value to use as delta angle in integral calculation )
- [in] fixed_face (index of a mesh polygon. Face rotation begin at this face)
- [out] verts (moved vertices)
- [out] edges (edges)
- [out] faces (faces)

rigid_origami_folding_screen_shot

Next is an example of the node tree.
rigid_origami_folding_node_tree_screen_shot

In this script, these classes are included.
- ObjectParams (Object wrapper)
- CreaseLines (Manager of crease lines)
- InsideVertex (Instance of each vertices inside of the target paper)
- FoldAngleCalculator (Calculator of each edge's fold angles)
- FaceRotation (Handler to operate face rotation)

Limitation:
- Some types of folding may not be done, which need some extra strength.
For example, https://www.youtube.com/watch?v=IP7WbV_yLgQ or https://www.foldfactory.com/offering/twist .
- It may be difficult to find where error occurred. For example, when using wrong edge list or opposite valleys/mountains.
- It's not always right shape because this calculation include differentiation and integration, but these are semi-pseudo implementation using only NumPy library.

## Preflight checklist

Put an x letter in each brackets when you're done this item:

- [x] Code changes complete.
- [] Code documentation complete.
- [x] Documentation for users complete (or not required, if user never sees these changes).
- [x ] Manual testing done. 
- [ ] Unit-tests implemented.
- [x] Ready for merge.

### CODE REVIEW

**Comment by asahidari** [comment]
These are some examples.

1. Twisted square folding.
rigid_origami_fold_example1

2. Water bomb tessellation no.1
rigid_origami_fold_example2

3. Water bomb tessellation no.2
rigid_origami_fold_example3

**Comment by vicdoval** [comment]
@asahidari  This is totally awesome!!! 
I wont be able to check the code in a couple of weeks in case anyone else want to check it before merging but I love the features this PR brings.

Also very interesting of the grasshopper example you provide is the mix with the kangaroo plugin that could be extrapolated to the Pulga Physics node.

Another idea @asahidari would be creating a full node instead of a SNL script :)  (maybe i will do it if you don't want to)

**Comment by asahidari** [comment]
I'm glad to hear that, and it's no problem to take some weeks. This has nearly 500 lines and may have some redundant codes.

Pulga Physics is fascinating, but cannot afford to understand it quickly when I writing this. If the origami node can collaborate the physics node, it is interesting and the tree node may become more simple. 

Currently, I think I do not have enough understanding sverchok's structure. So if I can easily write a full node and find where to place it, I want to do that. But if the collaboration with the physics node is not easy and there are some worries to collapse existing codes, please consider you would change that.

-----------
note about extra parameter:
step: I added this, because some folding process are different when this value changed. But it can be fixed value, like pi*0.05.
fixed face: Other faces are rotated following edge angles, but this face will not move as a start point of rotation. This also can be fixed, for example, at the center of the paper.

**Comment by enzyme69** [comment]
Maybe I just hit that merge button and test it to make crane origami... ☝️

**Comment by asahidari** [comment]
This node is currently deployed in nodes_scripts/SNLite_templates/utils directory. So you can test it with copying whole text of 'rigid_origami_folding.py' and paste it on Blender's text editor to use in SNLite script node.

But unfortunately, crane folding seems not to work well :( .
crane_strange_folding

It seems this script still has some bugs or limitations  (probably in face rotation process), because the motion is changed when the parameters 'step' and 'fixed_face' changed. I'll check it, but it might have good result when using with the physics node.

**Comment by asahidari** [comment]
> But unfortunately, crane folding seems not to work well :( .
> ...
> .... I'll check it, but it might have good result when using with the physics node.

As I checked my code and read the original article again, Crane folding in this way seems not to be successful. The author of the [original paper](https://origami.c.u-tokyo.ac.jp/~tachi/cg/SimulationOfRigidOrigami_tachi_4OSME.pdf) says in chapter.4:

`Many models can be folded to the desired state in this system, while many cannot be folded. ... For example, Miura fold, waterbomb tessellation, and pleated hypars are smoothly folded in the simulation, while a crane with finishing, four-crane base or tatou cannot be not folded to the final state.`

When you search for crane folding on line, [Origami Simulator](http://www.amandaghassaei.com/projects/origami_simulator/) can fold crane, but the app uses another method (using another type of crease line and pin-jointed truss model) and GPU to do more complicated calculation. And I could not find crane fold process using Rhino+Grasshopper, Houdini. So I think it is currently still tough to realize it.

**Comment by enzyme69** [comment]
Screen Shot 2020-07-27 at 11 37 53 am

I tested the node, very often when folding algorithm failled, it does not give error and no result.
https://youtu.be/X614Jv1kpX8

**Comment by asahidari** [comment]
> I tested the node, very often when folding algorithm failed, it does not give error and no result.
> ...
> I want to try a simple paper airplane, but somewhat so far cannot get result

I'm sorry that the node has some non-user-friendly interface. It's actually not easy to know where is wrong.

But it seems better to begin with simple fold patterns. Selecting crease lines (valley/mountain lines) randomly may confuse you and sometimes it actually cannot be foldable, because the angles of the edges around vertices inside strictly influences each other.

Searching for images with words like "crease line origami" may show you some fold patterns. As you know, crane and some patterns cannot be foldable now, but others (like Water bomb tessellations, Miura-fold, Yoshimura-fold and etc.) may show you the good folding process.

**Comment by nortikin** [comment]
Снимок экрана от 2020-07-27 22-29-12
```python
out_outer = []
out_inner = []
for i,v in enumerate(V1[0].data.vertices[:]):
    if 0 in [g.group for g in v.groups.values()]:
        out_outer.append(i)
    if 1 in [g.group for g in v.groups.values()]:
        out_inner.append(i)
append([out_outer,out_inner])
```
```python
out_outer = []
out_inner = []
for k,i in enumerate(V1[0]):
    if i[0] in V2[0][0] and i[1] in V2[0][0]:
        out_outer.append(k)
    if i[0] in V2[0][1] and i[1] in V2[0][1]:
        out_inner.append(k)
append([out_outer,out_inner])
```

**Comment by nortikin** [comment]
it is easily to rewrite in SNL and use without exec node

**Comment by nortikin** [comment]
Снимок экрана от 2020-07-27 23-07-15

**Comment by nortikin** [comment]
```python
outer = []
inner = []
for i,v in enumerate(object.data.vertices[:]):
    if 0 in [g.group for g in v.groups.values()]:
        outer.append(i)
    if 1 in [g.group for g in v.groups.values()]:
        inner.append(i)

mount = []
valley = []
for k,i in enumerate(object.data.edges[:]):
    if i.vertices[0] in outer and i.vertices[1] in outer:
        mount.append(i)
    if i.vertices[0] in inner and i.vertices[1] in inner:
        valley.append(i)
```

File: node_scripts/SNLite_templates/utils/rigid_origami_folding.py
Comment (nortikin) [inline node_scripts/SNLite_templates/utils/rigid_origami_folding.py:449]: here we can grub objects vertex groups

**Comment by enzyme69** [comment]
@nortikin Can you supply the blend of the paperplane? Or make video about it?

**Comment by enzyme69** [comment]
@nortikin Can you try?
https://parametrichouse.com/origami-world-2/

What's the sane way to do that?

I tried the SNL Rigid Origami on a Cylinder (not plane) and Blender frozen.

**Comment by enzyme69** [comment]
[paperplane_folding_study_002.blend.zip](https://github.com/nortikin/sverchok/files/4985656/paperplane_folding_study_002.blend.zip)


Not working @nortikin 
Screen Shot 2020-07-28 at 12 42 42 pm

**Comment by enzyme69** [comment]
Aha....... 😤 
Screen Shot 2020-07-28 at 12 59 48 pm

**Comment by enzyme69** [comment]
Screen Shot 2020-07-28 at 1 23 32 pm

**Comment by asahidari** [comment]
> solved with vertex groups
> two groups, you can decide what is valley and hill.
> ...

Thank you for the suggestions. It seems good idea. But I'm afraid some 'peripheral' edges might be included.
(In this example, edge (2, 3) is not crease line, but included.)
Screen Shot 2020-07-29 at 1 18 01

> Without mount/valley sockets. If we working with objects. But if we working with Sverchok mesh - it is another story.

Currently, this script is a SNLite script and that is why I put an 'object' to in-socket. But if this can be a full node and used with other nodes, it seems useful to use mesh instead of object.

**Comment by asahidari** [comment]
I just posted some folding examples in my blog. I hope these examples would be helpful to use this node.

http://asahidari.hatenablog.com/entry/2020/07/29/010410?_ga=2.66962478.1020466821.1595564630-1234164817.1595564630

**Comment by enzyme69** [comment]
Great blog post showing some of the interesting paper folding and its patterns!! 

Paper Crane is actually possible?

**Comment by nortikin** [comment]
@asahidari 
1.Did not found edges groups in blender 👎 (((
2.Agree, edge can be red as grouped. It could also included in both groups \o_O/
3.It works for me.
4.Case of inner sverchok data. We can insert some GP tricks with template, by the way.
5.Maybe to leave both approaches? object and sverchok.
6.No need to enter edges themselves, we can insert edges' indices. not to filter, less steps for user.
7.Thanks

**Comment by asahidari** [comment]
> @asahidari
> 1.Did not found edges groups in blender 👎 (((
> 2.Agree, edge can be red as grouped. It could also included in both groups \o_O/
> 3.It works for me.

This node simply ignores edges when they have a face only one side of them. (it assumes that edges inside of the paper must have faces both side of them.) But when error occurs, existence of these edges may be confusing or misleading.

> 4.Case of inner sverchok data. We can insert some GP tricks with template, by the way.

GP(Grease Pencil ?) seems to have many possibilities, but I cannot quickly get ideas how to use it because I'm not used to it :-(.

> 5.Maybe to leave both approaches? object and sverchok.
> 6.No need to enter edges themselves, we can insert edges' indices. not to filter, less steps for user.

Yes, the node does not need object socket now. Instead, vertices, edges and faces are needed to calculate angles and rotating faces.

Another idea:
Blender has [edge crease](https://docs.blender.org/manual/en/2.79/modeling/meshes/editing/edges.html#edge-crease) and the edge can have values between -1 to 1. When users set the crease values to all edges inside of the paper and the node uses the values (multiplying with math.pi) as angles, users will no longer need edge indices and various angles can be set. Instead, object socket will be needed again and the socket interface will be far different from original grasshopper node interface...

> 7.Thanks

Me too, thank you for insightful suggestions :-).

**Comment by asahidari** [comment]
> > 5.Maybe to leave both approaches? object and sverchok.
> > 6.No need to enter edges themselves, we can insert edges' indices. not to fi    lter, less steps for user.
>
> Yes, the node does not need object socket now. Instead, vertices, edges and fa    ces are needed to calculate angles and rotate faces.

Currently I'm changing the node's socket interface.
* Replacing object socket with verts/edges/faces sockets (following the previous suggestions)
* Adding crease-edge indices instead of edge themselves to input valley/mountain edges(following the previous suggestions)
* Adding crease-edge angles instead of a valley angle and a mountain angle. This change enable us to input some kind of angles. For example, water bomb square tessellation in attached image have 2 types of mountain angles (90 and 180 degr    ee.)

Input node will be simple like this.
rigid_origami_input_indices

Waterbomb_square_tessellation

Attached current script just in case.
[rigid_origami_folding_mesh_input.py.zip](https://github.com/nortikin/sverchok/files/5017200/rigid_origami_folding_mesh_input.py.zip)

**Comment by asahidari** [comment]
Also, I found a limitation (or a bug) only when a paper has multiple vertices inside.

When folding, the node calculates edge angles ('rho' angles in [original paper](https://origami.c.u-tokyo.ac.jp/~tachi/cg/SimulationOfRigidOrigami_tachi_4OSME.pdf)) around vertices inside, and multiply rotation-matrices in counter-clockwise order to get rotation angles. However, folding result will be sometimes changed or cannot be folded when the 'first' edge (in counter-clockwise order) changes. Currently, I'm trying to solve this, but have not been able to find any explanation in the source article.

rotation_edges_around_verticies_problem_modified

**Comment by vicdoval** [comment]
The interaction with the Pulga Physics would be:
Outputting from the node the force vector that is applied to each point and use it as initial velocity in the Pulga Physics. 
Probably it is too tricky for a regular user and you approach is simpler to understand and with better (and more accurate) results.

image

**Comment by vicdoval** [comment]
> So if I can easily write a full node and find where to place it, I want to do that

If you wish you can turn it in into a regular node. I  would recommend you to check a couple of PRs with new nodes as guide
https://github.com/nortikin/sverchok/pull/3263
https://github.com/nortikin/sverchok/pull/3264
https://github.com/nortikin/sverchok/pull/3284

**Comment by asahidari** [comment]
@vicdoval , Thanks for the supportive information.

> The interaction with the Pulga Physics would be:
> Outputting from the node the force vector that is applied to each point and use it as initial velocity in the Pulga Physics. 
> Probably it is too tricky for a regular user and you approach is simpler to understand and with better (and more accurate) results.

Initial-velocity vector to each point, I see. It seems the velocity should be given by crease edge rotation. It's not simple, but is cool if the motion is successfully generated. While making the full node (below), I'll try some implementation, but sorry if that could not make successful results.

if the vectors are simply 'verts_out - verts_in' of current node (attached image below), it is easy to implement and it works in some conditions, but it seems not to be the final solution, probably.
rigid_origami_folding_force_output

> If you wish you can turn it in into a regular node. I would recommend you to check a couple of PRs with new nodes as guide
> #3263
> #3264
> #3284

Thank you for the help. I can probably make the full node by following the guide. If the current issue would be solved, I'll try it. But please give me some more time (a couple of week, probably) if allowed.

**Comment by vicdoval** [comment]
@asahidari  take all the time you need. If anyone is in a hurry they can use the script you provided 😄

**Comment by asahidari** [comment]
> @asahidari take all the time you need. If anyone is in a hurry they can use the script you provided 😄

Thank you, @vicdoval , that's right and I feel relieved.

And good news. Fortunately, the problem I mentioned before (below) seems possible to be solved 👍 . 
> Also, I found a limitation (or a bug) only when a paper has multiple vertices inside.

After some more tests, I can start to go on next step (=creating a full node, write documents, etc.).

File: utils/rigid_origami_utils.py
Comment (Durman) [inline utils/rigid_origami_utils.py:107]:
```suggestion
    def generate_inside_vertices(cls, obj, crease_lines):
```
According Python convention

File: utils/rigid_origami_utils.py
Comment (Durman) [inline utils/rigid_origami_utils.py:112]:
```suggestion
        cls.indices = indices
```
It has matter in some cases. I think it would be more correct to use `cls` object.

File: utils/rigid_origami_utils.py
Comment (Durman) [inline utils/rigid_origami_utils.py:172]:
```suggestion
                if idx not in indices_outermost:
```
A bit better to read

File: nodes/modifier_change/rigid_origami.py
Comment (asahidari) [reply nodes/modifier_change/rigid_origami.py]: Thank you for the review :) . I use the way.

File: nodes/modifier_change/rigid_origami.py
Comment (asahidari) [reply nodes/modifier_change/rigid_origami.py]: Thank you for the suggestion. I use the way.

File: utils/rigid_origami_utils.py
Comment (asahidari) [reply utils/rigid_origami_utils.py]: Thank you :). I should have learned more about python before this ...

File: utils/rigid_origami_utils.py
Comment (asahidari) [reply utils/rigid_origami_utils.py]: Thank you. I'll check whether this method is currently used. If not used, I'll delete this method not to mislead someone later.

File: utils/rigid_origami_utils.py
Comment (asahidari) [reply utils/rigid_origami_utils.py]: Thank you. I'll check other function names whether they are defined in the same way.

File: utils/rigid_origami_utils.py
Comment (asahidari) [reply utils/rigid_origami_utils.py]: Thank you. I'll check whether any other class objects are used in the same way.

**Comment by asahidari** [comment]
Thank you for reviewing, @Durman , I learned a lot.

@Vicdoval , sorry for the waiting, but I think I've complete the origami full node.
Could you please confirm this can be merged ?

These are points I've worried whether I've done correctly.
- I gave up the collaboration with Pulga Physics node this time. I tried with output of initial velocity (=multiplying edge length with radian), but I could not get sound results. I'm sorry, but I attached a script node (not a full node) just in case. [rigid_origami_folding_with_physics.py.zip](https://github.com/nortikin/sverchok/files/5077981/rigid_origami_folding_with_physics.py.zip)
- I couldn't find any style guide for node icons. So following other node icons, I made an icon for origami node with Inkscape.
- Following the Pulga physics, I added this origami node to 'Alpha nodes' . Is this OK?
- When I made this node, I've found, read and followed these guidelines basically.
    - #3263 #3264 #3284 (You gave me in previous comment.)
    - http://nikitron.cc.ua/sverch/html/contributing.html
    - https://github.com/nortikin/sverchok/issues/2301
    - https://github.com/nortikin/sverchok/issues/1834

File: nodes/modifier_change/rigid_origami.py
Comment (vicdoval) [inline nodes/modifier_change/rigid_origami.py:27]: according to PEP8  

> Don't use spaces around the = sign when used to indicate a keyword argument, or when used to indicate a default value for an unannotated function parameter

so: 
```
name="Folding ratio"
description="Folding ratio from 0.0 to 1.0" 
```

and so on

File: utils/rigid_origami_utils.py
Comment (vicdoval) [inline utils/rigid_origami_utils.py:506]:
```suggestion
    def __rot_quat(cls, rad, n):
        n = n / np.linalg.norm(n)
        rad = float(rad)
        cos_rad = np.cos(rad)
        sin_rad = np.sin(rad)
        R = np.array([
                       [cos_rad + n[0] * n[0] * (1 - cos_rad), \
                       n[0] * n[1] * (1 - cos_rad) - n[2] * sin_rad , \
                       n[0] * n[2] * (1 - cos_rad) + n[1] * sin_rad , 0], \
                       [n[1] * n[0] * (1 - cos_rad)+ n[2] * sin_rad , \
                       cos_rad + n[1] * n[1] * (1-cos_rad ), \
                       n[1] * n[2] * (1 - cos_rad) - n[0] * sin_rad , 0], \
                       [n[2] * n[0] * (1 - cos_rad) - n[1] * sin_rad , \
                       n[2] * n[1] * (1 - cos_rad) + n[0] * sin_rad , \
                       cos_rad + n[2] * n[2] * (1 - cos_rad), 0], \
                       [0, 0, 0, 1]])
        return R
```

Comment (asahidari) [reply utils/rigid_origami_utils.py]: Thank you. It had not been easy to read, but your modification made it really simple and readable!

**Comment by vicdoval** [comment]
image

@asahidari  Thanks again and congratulations!
Aside of the couple of format suggestions i mention in the code everything seems pretty clear and well documented.

Don't worry about the interaction with the Pulga node, it was a side idea, also Pulga Physics is pretty experimental and it may change in the future that is why it is in the Alpha Category, if your node is stable, has documentation and it is in its final state maybe it would be better to move it to Modifier Make category or Modify Change (it could work in both).

Maybe would be nice to add one or more examples (exported jsons) to the examples folder in the category you find more suitable for them :)

File: nodes/modifier_change/rigid_origami.py
Comment (asahidari) [reply nodes/modifier_change/rigid_origami.py]: Thank you for the suggestion, and sorry that I had missed it. I modified the places which has the same issue at https://github.com/nortikin/sverchok/pull/3416/commits/2d0560d679e47680fad664d99c79d1b42ceb8880 .

**Comment by asahidari** [comment]
> @asahidari Thanks again and congratulations!
> Aside of the couple of format suggestions i mention in the code everything seems pretty clear and well documented.

Thank you @vicdoval, I appreciated your review. I was relieved to hear that. (And the image you attached is photorealistic and very cool!  😄)

> Don't worry about the interaction with the Pulga node, it was a side idea, also Pulga Physics is pretty experimental and it may change in the future that is why it is in the Alpha Category, if your node is stable, has documentation and it is in its final state maybe it would be better to move it to Modifier Make category or Modify Change (it could work in both).

OK, I agree with you about the physics node. And I just moved it from Alpha Category to Modify Change at https://github.com/nortikin/sverchok/pull/3416/commits/d120e5c110f356380eee8f3dd938d7465417b3c2 .
 
> Maybe would be nice to add one or more examples (exported jsons) to the examples folder in the category you find more suitable for them :)

I agree it is good for users to import examples from json files. However, source objects cannot be exported and the edge indices might be different between users. So, I displayed source objects and the edge indices in example images instead. Is it OK without exported json files in this case? Anyway, I added another example at https://github.com/nortikin/sverchok/pull/3416/commits/149e179485407959a4717a2c345e444b607fb156 .

**Comment by asahidari** [comment]
> > Maybe would be nice to add one or more examples (exported jsons) to the examples folder in the category you find more suitable for them :)

> I agree it is good for users to import examples from json files. However, currently source objects are not exported together and the edge indices might be different between users. So, I displayed source objects and the edge indices in example images instead. Is it OK without exported json files in this case?

Sorry, I did not know Text In/Out node can dump/load object mesh datas. (These are very useful! I found that when I try other json_examples. And even without these nodes, simple data can be made with vector/edge indices list/polygon indices list.)

I added two examples at https://github.com/nortikin/sverchok/pull/3416/commits/6eea52089e629cf4c835b1e0b5fe1d106b343401 . Could you check these once more?

**Comment by vicdoval** [comment]
@asahidari  Im glad you found the way to add some examples, they are awesome.

The PR is almost ready 😄 , the only thing is that there is a duplicated example:
Water bomb origami tessellation and Water bomb origami tessellation square

**Comment by asahidari** [comment]
> @asahidari Im glad you found the way to add some examples, they are awesome.
> 
> The PR is almost ready 😄 , the only thing is that there is a duplicated example:
> Water bomb origami tessellation and Water bomb origami tessellation square

Thanks for the pointing out, and I'm sorry. Adding examples is backfired... All I wanted to do has been completed, and will do nothing if there are no need. Really, sorry 😰.

**Review by nortikin — APPROVED** [review]

**Comment by vicdoval** [comment]
@asahidari  Thanks and congratulations for your first full featured node in Sverchok!!

**Comment by nortikin** [comment]
lovely

**Comment by asahidari** [comment]
This experience had been very exciting!

@vicdoval , thank you for many supports and advices!

@nortikin , thank you for many suggestions and approving this PR!

@enzyme69 , thank you for trying this node and suggesting a lot!

@Durman , thank you for reviewing and many advices!

**Comment by asahidari** [comment]
@nortikin 
I have to apologize to you. While I was involved in this PR, I have missed all mails from GitHub (I noticed those mails some days after this PR was merged!). Some were important and I probably missed some opportunities. Despite my this rude attitude, however, you have involved in this PR. I sincerely appreciated, thanks.

**Comment by mavizasyon** [comment]
How can I do this flasher origami? I have zero knowledge about sverchok nodes and tutorials are seem to outdated. Red mountain, blue valley
Eskiz1
[flasher.zip](https://github.com/user-attachments/files/19651485/flasher.zip)

**Comment by asahidari** [comment]
@mavizasyon

Would you try to see the attached file? Or follow the steps bellow (, images below might help you understand):

1. Fill the center hole in advance to be used as a fixed face for 'Rigid origami' node parameter.
2. Use 'Viewer index+' node to see all edge indices of your polygon.
3. Create list of mountain-edge and valley-edge indices with the final angles of all edges.
4. Feed the parameters for 'Rigid origami' node with connecting nodes and 'Viewer Draw' node to see the result.
5. Change 'Folding ratio' parameter from 0.0 to 1.0 to see the folding step.

I tried it and saw some results. When 'Folding ratio' in 'Rigid origami' node is from 0.0 to 0.3, the result seems well. But when the ratio is over 0.4, the output were not so good and some adjustments (changing final angles, etc.) might be needed.

I hope this comment will help you.

Screen Shot 2025-04-09 at 21 10 50
Screen Shot 2025-04-09 at 21 20 40
Screen Shot 2025-04-09 at 21 26 20
[flasher_modified.blend.zip](https://github.com/user-attachments/files/19667886/flasher_modified.blend.zip)

**Comment by mavizasyon** [comment]
@asahidari 

could you help me to create shrinking cylinder origami? I have tried but it freezes a lot and couldn't make it work perfectly, sometimes nothing happens sometimes gives errors. It is hard look and write edge indicies. Is there select and create list option? And another question is how can I achieve self collison like origami simulator.

image

Files;

https://drive.google.com/file/d/1E1F5y1pzQrXib7RH0HmVs2QX0z9YHC6v/view?usp=sharing
https://drive.google.com/file/d/1C2vfBMPwy3D3Xxl9FGQM7yfOxGgeqxwz/view?usp=sharing

**Comment by asahidari** [comment]
@mavizasyon

I'm so sorry to be late for replying. And unfortunately, I do not have any environments to check your cylinder now. So, I can only speculate about your issue, but I try to answer.

* Try to make shorter size cylinders at first. The cylinder, in the image you posted, seems to have already eight sections in vertical direction. If the model has more complicated structure, the more difficult to find the problem. So it seems better to start with simple models to find the solution.

* Check the surface direction (back or front) of each faces of your model. If some faces have different direction, they might cause problems.

If these advice will not work well, I cannot find any solutions right now. And the problem might be caused by the limitation of this node(, because this node is for "rigid" origami, not for lenient, pliable origami).

For the other question you posted: unfortunately, this node is currently not capable of self collision detection. This node is not based on Origami Simulator, because the simulator uses GPU.

**Pull request merged**
